src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs

```diff
@@ -1480,8 +1486,31 @@ private async Task<Solution> ComputeFrozenSolutionAsync(CancellationToken cancel
         /// </summary>
         internal Solution WithFrozenPartialCompilationIncludingSpecificDocument(DocumentId documentId, CancellationToken cancellationToken)
         {
-            var newCompilationState = this.CompilationState.WithFrozenPartialCompilationIncludingSpecificDocument(documentId, cancellationToken);
-            return new Solution(newCompilationState);
+            AsyncLazy<Solution>? lazySolution;
+            lock (_documentIdToFrozenSolution)
+            {
+                if (!_documentIdToFrozenSolution.TryGetValue(documentId, out lazySolution))
+                {
+                    // in a local function to prevent lamdba allocations when not needed.
+                    lazySolution = CreateLazyFrozenSolution(this.CompilationState, documentId);
+                    _documentIdToFrozenSolution.Add(documentId, lazySolution);
+                }
+            }
+
+            return lazySolution.GetValue(cancellationToken);
+
+            static AsyncLazy<Solution> CreateLazyFrozenSolution(SolutionCompilationState compilationState, DocumentId documentId)
+            {
+                return new AsyncLazy<Solution>(
+                    cancellationToken => Task.FromResult(ComputeFrozenSolution(compilationState, documentId, cancellationToken)),
+                    cancellationToken => ComputeFrozenSolution(compilationState, documentId, cancellationToken));
```

Comment:
a bit of an odd pattern that's been arising recently.  All the actual computation work is async, but we're storing it in an AsyncLazy.  This is so that we have the nice benefit that N callers can be trying to get the value at the same time, but only one will compute it.  Also, if all of htem cancel, the underlying work cancels.  (as opposed to if N callers are asking, and the one computing ends up cancelling, causing the next that is waiting to have to redo all the work).